File: app/src/main/java/com/bopr/android/smailer/CallProcessorService.java
package com.bopr.android.smailer;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static java.util.Objects.requireNonNull;

/**
 * Service that processes phone event.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class CallProcessorService extends IntentService {

    private static Logger log = LoggerFactory.getLogger("CallProcessorService");

    private static final String EXTRA_EVENT = "event";

    public CallProcessorService() {
        super("call-processor");
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        log.debug("Running");

        PhoneEvent event = requireNonNull(requireNonNull(intent).getParcelableExtra(EXTRA_EVENT));
        try (Database database = new Database(this)) {
            new CallProcessor(this, database).process(event);
        }
    }

    /**
     * Start the service.
     *
     * @param context context
     * @param event   event
     */
    public static void startCallProcessingService(@NonNull Context context, @NonNull PhoneEvent event) {
        log.debug("Starting service for: " + event);

        context.startService(new Intent(context, CallProcessorService.class)
                .putExtra(EXTRA_EVENT, event));
    }
}

File: app/src/main/java/com/bopr/android/smailer/CallProcessorService.kt
package com.bopr.android.smailer

import android.app.IntentService
import android.content.Context
import android.content.Intent
import org.slf4j.LoggerFactory

/**
 * Service that processes phone event.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
class CallProcessorService : IntentService("call-processor") {

    private lateinit var callProcessor: CallProcessor
    private lateinit var database: Database

    override fun onCreate() {
        super.onCreate()
        database = Database(this)
        callProcessor = CallProcessor(this, database)
    }

    override fun onDestroy() {
        database.close()
        super.onDestroy()

        log.debug("Destroyed")
    }

    override fun onHandleIntent(intent: Intent?) {
        log.debug("Running")
        intent?.getParcelableExtra<PhoneEvent>(EXTRA_EVENT)?.let {
            callProcessor.process(it)
        }
    }

    companion object {

        private val log = LoggerFactory.getLogger("CallProcessorService")
        private const val EXTRA_EVENT = "event"

        /**
         * Start the service.
         *
         * @param context context
         * @param event   event
         */
        fun startCallProcessingService(context: Context, event: PhoneEvent) {
            log.debug("Starting service for: $event")
            context.startService(Intent(context, CallProcessorService::class.java)
                    .putExtra(EXTRA_EVENT, event))
        }
    }
}